File: Python/Minimum Add to Make Parentheses Valid.py
class Solution:
    """
    Time:   O(n)
    Memory: O(1)
    """

    def minAddToMakeValid(self, s: str) -> int:
        score = add = 0

        for ch in s:
            score += -1 if ch == ')' else 1
            if score < 0:
                add += 1
                score = 0

        return score + add

File: Java/Minimum Add to Make Parentheses Valid.java
public class Solution {

    /**
     * Time:   O(n)
     * Memory: O(1)
     */
    public int minAddToMakeValid(String s) {
        int add = 0, score = 0;

        for (int i = 0; i < s.length(); ++i) {
            score += s.charAt(i) == '(' ? 1 : -1;
            if (score == -1) {
                ++add;
                ++score;
            }
        }

        return add + score;
    }
}
